programs/castle-vault/src/asset_container/reserves.rs

```diff
@@ -67,13 +74,15 @@ impl AssetContainer<Reserves> {
 
     pub fn get_apr(
         &self,
-        weights: &dyn Index<Provider, Output = Rate>,
-        allocations: &dyn Index<Provider, Output = u64>,
+        weights: &dyn Index<Provider, Output = Option<Rate>>,
+        allocations: &dyn Index<Provider, Output = Option<u64>>,
     ) -> Result<Rate, ProgramError> {
         self.into_iter()
+            .filter(|(_, r)| !r.is_none())
+            .map(|(p, r)| (p, r.unwrap()))
             .map(|(p, r)| {
-                r.calculate_return(allocations[p])
-                    .and_then(|r| weights[p].try_mul(r))
+                r.calculate_return(allocations[p].unwrap())
```

Comment:
In general, `unwrap` is bad practice since it doesn't give a very useful error message and can mask possible errors from downstream users.

Better to rewrite like so:
```
        self.into_iter()
            .map(|(p, o)| (o, allocations[p], weights[p]))
            .flat_map(|t| match t {
                (Some(r), Some(a), Some(w)) => Some((r, a, w)),
                _ => None,
            })
            .map(|(r, a, w)| r.calculate_return(a).and_then(|ret| w.try_mul(ret)))
            .try_fold(Rate::zero(), |acc, r| acc.try_add(r?))
```